File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/eal_to_amap.py
from collections import defaultdict

def eal_to_amap(eal):
    # Khoi tao ban do ke 2 chieu voi gia tri mac dinh la 0
    amap = defaultdict(lambda: defaultdict(int))

    for u, v, w in eal:
        amap[u][v] = w  # Gan so canh tu u den v la w

    return amap

# Nhap tu ban phim
n = int(input("Nhap so dinh: "))
k = int(input("Nhap so bo (u v w): "))

eal = []
print("Nhap cac bo (u v w):")
for _ in range(k):
    u, v, w = map(int, input().split())
    eal.append((u, v, w))

amap = eal_to_amap(eal)

print("Ban do ke (AMap):")
for u in range(n):
    print(f"{u}:", end=" ")
    for v in amap[u]:
        print(f"({v},{amap[u][v]})", end=" ")
    print()

File: combinatorics/resource/ST/project_5/bai_toan_4/mutil_graph/eal_to_amap.cpp
#include <iostream>
#include <map>
#include <vector>
#include <tuple>

using namespace std;

// Ham chuyen EAL sang AMap
map<int, map<int, int>> ealToAmap(const vector<tuple<int, int, int>>& eal) {
    map<int, map<int, int>> amap;

    for (auto [u, v, w] : eal) {
        amap[u][v] = w;  // Gan so canh tu u den v la w
    }

    return amap;
}

int main() {
    int n, k;
    cout << "Nhap so dinh: ";
    cin >> n;
    cout << "Nhap so bo (u v w): ";
    cin >> k;

    vector<tuple<int, int, int>> eal;
    cout << "Nhap cac bo (u v w):\n";
    for (int i = 0; i < k; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        eal.emplace_back(u, v, w);
    }

    auto amap = ealToAmap(eal);

    cout << "Ban do ke (AMap):\n";
    for (int u = 0; u < n; ++u) {
        cout << u << ": ";
        for (auto& [v, w] : amap[u]) {
            cout << "(" << v << "," << w << ") ";
        }
        cout << endl;
    }

    return 0;
}